doc/man/man1/ldapvc.1
.TH LDAPVC 1 "RELEASEDATE" "OpenLDAP LDVERSION"
.\" $OpenLDAP$
.\" Copyright 1998-2026 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.SH NAME
ldapvc \- LDAP verify credentials tool
.SH SYNOPSIS
.B ldapvc
[\c
.BR \-V [ V ]]
[\c
.BI \-d \ debuglevel\fR]
[\c
.BR \-a ]
[\c
.BR \-b ]
[\c
.BR \-n ]
[\c
.BR \-v ]
[\c
.BR \-x ]
[\c
.BI \-D \ binddn\fR]
[\c
.BR \-W ]
[\c
.BI \-w \ passwd\fR]
[\c
.BI \-y \ passwdfile\fR]
[\c
.BI \-H \ ldapuri\fR]
[\c
.BR \-e \ [ ! ] \fIext\fP [ =\fIextparam\fP ]]
[\c
.BR \-E \ [ ! ] \fIext\fP [ =\fIextparam\fP ]]
[\c
.BI \-o \ opt \fR[= optparam \fR]]
[\c
.BI \-O \ security-properties\fR]
[\c
.BR \-I ]
[\c
.BR \-Q ]
[\c
.BR \-N ]
[\c
.BI \-U \ authcid\fR]
[\c
.BI \-R \ realm\fR]
[\c
.BI \-X \ authzid\fR]
[\c
.BI \-Y \ mech\fR]
[\c
.BR \-Z [ Z ]]
\c
.I Distinguished Name \
\c
.I [Credentials]
.SH DESCRIPTION
.I ldapvc
implements the LDAP "Verify Credentials" extended operation.
.LP
.B Verify Credentials
operation behaves like LDAP Bind but has no impact upon the underlying LDAP session.
.SH OPTIONS
.TP
.BR \-V [ V ]
Print version info.
If \fB\-VV\fP is given, only the version information is printed.
.TP
.BI \-d \ debuglevel
Set the LDAP debugging level to \fIdebuglevel\fP.
.B ldapvc
must be compiled with LDAP_DEBUG defined for this option to have any effect.
.TP
.B \-a
Print the authzID resulting from a successful verification of credentials.
.TP
.B \-b
Print the results from the ppolicy control after verification of credentials.
.TP
.B \-n
Show what would be done, but don't actually perform the operation.
Useful for
debugging in conjunction with \fB\-v\fP.
.TP
.B \-v
Run in verbose mode, with many diagnostics written to standard output.
.TP
.B \-x 
Use simple authentication instead of SASL.
.TP
.BI \-D \ binddn
Use the Distinguished Name \fIbinddn\fP to bind to the LDAP directory.
For SASL binds, the server is expected to ignore this value.
.TP
.B \-W
Prompt for simple authentication.
This is used instead of specifying the password on the command line.
.TP
.BI \-w \ passwd
Use \fIpasswd\fP as the password for simple authentication.
.TP
.BI \-y \ passwdfile
Use complete contents of \fIpasswdfile\fP as the password for
simple authentication.
.TP
.BI \-H \ ldapuri
Specify URI(s) referring to the ldap server(s); only the protocol/host/port
fields are allowed; a list of URI, separated by whitespace or commas
is expected.
.TP
.BR \-e \ [ ! ] \fIext\fP [ =\fIextparam\fP ]
.TP
.BR \-E \ [ ! ] \fIext\fP [ =\fIextparam\fP ]

Specify general extensions with \fB\-e\fP and Verify Credentials extensions with \fB\-E\fP.
\'\fB!\fP\' indicates criticality.

General extensions:
.nf
  [!]assert=<filter>    (an RFC 4515 Filter)
  [!]bauthzid           (RFC 3829 authzid control)
  [!]chaining[=<resolve>[/<cont>]]
  [!]manageDSAit
  [!]noop
  ppolicy
  [!]postread[=<attrs>] (a comma-separated attribute list)
  [!]preread[=<attrs>]  (a comma-separated attribute list)
  [!]relax
  sessiontracking[=<username>]
  abandon,cancel,ignore (SIGINT sends abandon/cancel,
  or ignores response; if critical, doesn't wait for SIGINT.
  not really controls)
.fi
.sp
Verify Credentials extensions:
.sp
The following options set SASL params on the Verify Credentials request:
.nf
  authcid=<authcid>    (SASL Authentication Identity "dn:<dn>" or "u:<user>")
  authzid=<authzid>    (SASL Authorization Identity "dn:<dn>" or "u:<user>")
  mech=<mech>          (SASL mechanism default e.g. Simple)
  realm=<realm>        (SASL Realm, defaults to none)
  sasl=a[utomatic]|i[nteractive]|q[uiet]  (SASL mode defaults to automatic if any other -E option provided, otherwise none)
  secprops=<secprops>  (SASL Security Properties)
.fi
.TP
.BI \-o \ opt \fR[= optparam \fR]

Specify any
.BR ldap.conf (5)
option or one of the following:
.nf
  nettimeout=<timeout>  (in seconds, or "none" or "max")
  ldif_wrap=<width>     (in columns, or "no" for no wrapping)
.fi

.B -o
option that can be passed here, check
.BR ldap.conf (5)
for details.
.TP
.BI \-O \ security-properties
Specify SASL security properties.
.TP
.B \-I
Enable SASL Interactive mode.  Always prompt.  Default is to prompt
only as needed.
.TP
.B \-Q
Enable SASL Quiet mode.  Never prompt.
.TP
.B \-N
Do not use reverse DNS to canonicalize SASL host name.
.TP
.BI \-U \ authcid
Specify the authentication ID for SASL bind. The form of the ID
depends on the actual SASL mechanism used.
.TP
.BI \-R \ realm
Specify the realm of authentication ID for SASL bind. The form of the realm
depends on the actual SASL mechanism used.
.TP
.BI \-X \ authzid
Specify the requested authorization ID for SASL bind.
.I authzid
must be one of the following formats:
.BI dn: "<distinguished name>"
or
.BI u: <username>
.TP
.BI \-Y \ mech
Specify the SASL mechanism to be used for authentication. If it's not
specified, the program will choose the best mechanism the server knows.
.TP
.BR \-Z [ Z ]
Issue StartTLS (Transport Layer Security) extended operation. If you use
\fB\-ZZ\fP, the command will require the operation to be successful.
.SH EXAMPLE
.nf
    ldapvc \-x "uid=Alice,ou=People,dc=example,dc=com"
.fi
.SH "SEE ALSO"
.BR ldap.conf (5),
.BR ldap (3),
.BR ldap_extended_operation (3)
.SH AUTHOR
The OpenLDAP Project <http://www.openldap.org/>
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.6 Change Log

OpenLDAP 2.6.13 Engineering

OpenLDAP 2.6.12 Release (2026/01/29)
	Fixed libldap to reject empty types in LDIF (ITS#10429)
	Fixed libldap to not scroll past nul bytes (ITS#10430)
	Fixed libldap to enforce stop when encountering nul-leading line (ITS#10431)

OpenLDAP 2.6.11 Release (2026/01/28)
	Fixed slapd to use fresh timestamp for lastbind (ITS#10379)
	Fixed slapd delta-syncrepl to always use logDB rootdn (ITS#10360)
	Fixed slapd reverse lookup of proxied IPv6 addresses (ITS#10387)
	Fixed slapd logging buffer overflow (ITS#10410)
	Fixed slapd-ldap response when invalid secprops is configured (ITS#10392)
	Fixed slapd-mdb error when deleting last child of a branch (ITS#10304)
	Fixed slapd-mdb check for pool pause in search (ITS#10191)
	Fixed slapo-memberof clash with refint on subtree rename (ITS#10398)
	Fixed slapo-syncprov use correct rootDN for accesslog replay (ITS#10385)
	Minor Cleanup
		ITS#7901
		ITS#10329
		ITS#10335
		ITS#10339
		ITS#10343
		ITS#10344
		ITS#10345
		ITS#10347
		ITS#10348
		ITS#10349
		ITS#10353
		ITS#10358
		ITS#10359
		ITS#10366
		ITS#10367
		ITS#10369
		ITS#10370
		ITS#10371
		ITS#10372
		ITS#10374
		ITS#10375
		ITS#10376
		ITS#10377
		ITS#10379
		ITS#10380
		ITS#10381
		ITS#10384
		ITS#10388
		ITS#10390
		ITS#10391
		ITS#10400
		ITS#10401
		ITS#10408

OpenLDAP 2.6.10 Release (2025/05/22)
	Added slapd microsecond timestamp format for local logging (ITS#10140)
	Fixed libldap ldap_result behavior with LDAP_MSG_RECEIVED (ITS#10229)
	Fixed lloadd handling of starttls critical (ITS#10323)
	Fixed slapd syncrepl when used with slapo-rwm (ITS#10290)
	Fixed slapd regression with certain searches (ITS#10307)
	Fixed slapo-autoca olcAutoCAserverClass object (ITS#10288)
	Fixed slapo-pcache caching behaviors (ITS#10270)
	Minor Cleanup
		ITS#7080
		ITS#7249
		ITS#9934
		ITS#10020
		ITS#10168
		ITS#10226
		ITS#10279
		ITS#10299
		ITS#10302
		ITS#10309
		ITS#10312
		ITS#10320
		ITS#10325
		ITS#10327
		ITS#10328
		ITS#10331
		ITS#10336

OpenLDAP 2.6.9 Release (2024/11/26)
	Fixed libldap TLS connection timeout handling (ITS#8047)
	Fixed libldap GnuTLS incompatible pointer type (ITS#10253)
	Fixed libldap OpenSSL set_ciphersuite error handling (ITS#10223)
	Fixed libldap to check for OpenSSL EVP_Digest* failure (ITS#10224)
	Fixed slapd cn=config disallowed modification of cn=schema (ITS#10256)
	Fixed slapd syncrepl assert during refresh at shutdown (ITS#10232)
	Fixed slapd syncrepl retry state during refreshDone (ITS#10234)
	Fixed slapd-ldap use of multi-precision add for op counters (ITS#10237)
	Fixed slapd-mdb idl intersection (ITS#10233)
	Fixed slapd-wt idl intersection (ITS#10233)
	Fixed slapo-memberof to omit dynamic values (ITS#10230)
	Fixed slapo-nestgroup leak in nestgroup_memberFilter (ITS#10249)
	Fixed slapo-translucent regression with subordinate databases (ITS#10248)
	Fixed slapo-translucent regression when requesting attributes (ITS#10272)
	Fixed slappw-argon2 defaults to be more secure (ITS#9827)
	Minor Cleanup
		ITS#10155
		ITS#10218
		ITS#10219
		ITS#10227
		ITS#10231
		ITS#10235
		ITS#10263
		ITS#10264

OpenLDAP 2.6.8 Release (2024/05/21)
	Fixed libldap exit handling with OpenSSL3 again (ITS#9952)
	Fixed libldap OpenSSL channel binding digest (ITS#10216)
	Fixed slapd handling of large uid/gids peercred auth (ITS#10211)
	Fixed slapd-asyncmeta/meta target structure allocations (ITS#10197)
	Fixed slapd-meta with dynlist (ITS#10164)
	Fixed slapd-meta binds when proxying internal op (ITS#10165)
	Added slapo-nestgroup overlay (ITS#10161)
	Added slapo-memberof 'addcheck' option (ITS#10167)
	Fixed slapo-accesslog startup initialization (ITS#10170)
	Fixed slapo-constraint double free on invalid attr (ITS#10204)
	Fixed slapo-dynlist with abandoned operations (ITS#10044)
	Build
		Fixed build with gcc14.x (ITS#10166)
		Fixed back-perl with clang15 (ITS#10177)
		Fixed to reduce systemd dependencies (ITS#10214)
	Contrib
		Added slapo-alias contrib module (ITS#10104, ITS#10182)
		Fixed slapo-autogroup to work with slapo-dynlist (ITS#10185)
		Fixed smbk5pwd implicit function declaration (ITS#10206)
	Documentation
		Fixed slapo-memberof exattr requirements (ITS#7400)
		Fixed slapo-memberof is no longer deprecated (ITS#7400)
	Minor Cleanup
		ITS#9921
		ITS#10103
		ITS#10171
		ITS#10172
		ITS#10173
		ITS#10179
		ITS#10183
		ITS#10186
		ITS#10188
		ITS#10193
		ITS#10209

OpenLDAP 2.6.7 Release (2024/01/29)
	Added slapo-dynlist option to disable filter support (ITS#10025)
	Fixed liblber missing newline on long msg (ITS#10105)
	Fixed libldap exit handling with OpenSSL3 (ITS#9952)
	Fixed libldap with TLS and multiple ldap URIs (ITS#10101)
	Fixed libldap OpenSSL cipher suite handling (ITS#10094)
	Fixed libldap OpenSSL 3.0 and Diffie-Hellman param files (ITS#10124)
	Fixed libldap timestamps on Windows (ITS#10100)
	Fixed lloadd to work when resolv.conf is missing (ITS#10070)
	Fixed lloadd handling of closing connection (ITS#10083)
	Fixed lloadd tiers to be correctly linked on startup (ITS#10142)
	Fixed slapd to honour disclose in matchedDN handling (ITS#10139)
	Fixed slapd handling of regex testing in ACLs (ITS#10089)
	Fixed slapd sync replication with glued database (ITS#10080)
	Fixed slapd local logging on Windows (ITS#10092)
	Fixed slapd-asyncmeta when remote suffix is empty (ITS#10076)
	Fixed slapo-dynlist so it can't be global (ITS#10091)
	Build
		Fixed lloadd type mismatches (ITS#10074)
		Fixed builds for Windows (ITS#10117)
		Fixed build with clang16 (ITS#10123)
	Documentation
		Fixed slapo-homedir(5) attribute name for olcHomedirArchivePath (ITS#10057)
	Minor Cleanup
		ITS#10059
		ITS#10068
		ITS#10098
		ITS#10109
		ITS#10110
		ITS#10129
		ITS#10130
		ITS#10135
		ITS#10143
		ITS#10144
		ITS#10145
		ITS#10153

OpenLDAP 2.6.6 Release (2023/07/31)
	Fixed slapd cn=config incorrect handling of paused (ITS#10045)
	Fixed slapd-meta to account for MOD ops being optional (ITS#10067)
	Fixed slapd-asyncmeta to account for MOD ops being optional (ITS#10067)


clients/tools/ldapvc.c
/* ldapvc.c -- a tool for verifying credentials */
/* $OpenLDAP$ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 1998-2026 The OpenLDAP Foundation.
 * Portions Copyright 2010 Kurt D. Zeilenga.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in the file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* Portions Copyright (c) 1992-1996 Regents of the University of Michigan.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that this notice is preserved and that due credit is given
 * to the University of Michigan at Ann Arbor.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.  This
 * software is provided ``as is'' without express or implied warranty.
 */
/* ACKNOWLEDGEMENTS:
 * This work was originally developed by Kurt D. Zeilenga for inclusion
 * in OpenLDAP Software based, in part, on other client tools.
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/socket.h>
#include <ac/string.h>
#include <ac/time.h>
#include <ac/unistd.h>

#include <ldap.h>
#include "lutil.h"
#include "lutil_ldap.h"
#include "ldap_defaults.h"

#include "common.h"

static int req_authzid = 0;
static int req_pp = 0;

#if defined(LDAP_API_FEATURES_VERIFY_CREDENTIALS_INTERACTIVE) && defined(HAVE_CYRUS_SASL)
#define LDAP_SASL_NONE (~0U)
static unsigned vc_sasl = LDAP_SASL_NONE;
static char *vc_sasl_realm = NULL;
static char *vc_sasl_authcid = NULL;
static char *vc_sasl_authzid = NULL;
static char *vc_sasl_mech = NULL;
static char *vc_sasl_secprops = NULL;
#endif
static char * dn = NULL;
static struct berval cred = {0, NULL};

void
usage( void )
{
	fprintf( stderr, _("Issue LDAP Verify Credentials operation to verify a user's credentials\n\n"));
	fprintf( stderr, _("usage: %s [options] [DN [cred]])\n"), prog);
	fprintf( stderr, _("where:\n"));
	fprintf( stderr, _("    DN\tDistinguished Name\n"));
	fprintf( stderr, _("    cred\tCredentials (prompt if not present)\n"));
	fprintf( stderr, _("options:\n"));
	fprintf( stderr, _("    -a\tRequest AuthzId\n"));
	fprintf( stderr, _("    -b\tRequest Password Policy Information\n"));
	fprintf( stderr, _("    -E sasl=(a[utomatic]|i[nteractive]|q[uiet]>\tSASL mode (defaults to automatic if any other -E option provided, otherwise none))\n"));
	fprintf( stderr, _("    -E mech=<mech>\tSASL mechanism (default "" e.g. Simple)\n"));
	fprintf( stderr, _("    -E realm=<realm>\tSASL Realm (defaults to none)\n"));
	fprintf( stderr, _("    -E authcid=<authcid>\tSASL Authentication Identity (defaults to USER)\n"));
	fprintf( stderr, _("    -E authzid=<authzid>\tSASL Authorization Identity (defaults to none)\n"));
	fprintf( stderr, _("    -E secprops=<secprops>\tSASL Security Properties (defaults to none)\n"));
	tool_common_usage();
	exit( EXIT_FAILURE );
}


const char options[] = "abE:"
	"d:D:e:H:InNO:o:QR:U:vVw:WxX:y:Y:Z";

int
handle_private_option( int i )
{
	switch ( i ) {
		char	*control, *cvalue;
	case 'E': /* vc extension */
		if( protocol == LDAP_VERSION2 ) {
			fprintf( stderr, _("%s: -E incompatible with LDAPv%d\n"),
				prog, protocol );
			exit( EXIT_FAILURE );
		}

		/* should be extended to support comma separated list of
		 *	[!]key[=value] parameters, e.g.  -E !foo,bar=567
		 */

		cvalue = NULL;
		if( optarg[0] == '!' ) {
			optarg++;
		}

		control = optarg;
		if ( (cvalue = strchr( control, '=' )) != NULL ) {
			*cvalue++ = '\0';
		}

		if (strcasecmp(control, "sasl") == 0) {
#if defined(LDAP_API_FEATURES_VERIFY_CREDENTIALS_INTERACTIVE) && defined(HAVE_CYRUS_SASL)
			if (vc_sasl != LDAP_SASL_NONE) {
				fprintf(stderr,
				    _("SASL option previously specified\n"));
				exit(EXIT_FAILURE);
			}
			if (cvalue == NULL) {
				fprintf(stderr,
					_("missing mode in SASL option\n"));
				exit(EXIT_FAILURE);
			}

			switch (*cvalue) {
			case 'a':
			case 'A':
				vc_sasl = LDAP_SASL_AUTOMATIC;
				break;
			case 'i':
			case 'I':
				vc_sasl = LDAP_SASL_INTERACTIVE;
				break;
			case 'q':
			case 'Q':
				vc_sasl = LDAP_SASL_QUIET;
				break;
			default:
				fprintf(stderr,
					_("unknown mode %s in SASL option\n"), cvalue);
				exit(EXIT_FAILURE);
			}
#else
			fprintf(stderr,
				_("%s: not compiled with SASL support\n"), prog);
			exit(EXIT_FAILURE);
#endif

		} else if (strcasecmp(control, "mech") == 0) {
#if defined(LDAP_API_FEATURES_VERIFY_CREDENTIALS_INTERACTIVE) && defined(HAVE_CYRUS_SASL)
			if (vc_sasl_mech) {
				fprintf(stderr,
				    _("SASL mech previously specified\n"));
				exit(EXIT_FAILURE);
			}
			if (cvalue == NULL) {
				fprintf(stderr,
					_("missing mech in SASL option\n"));
				exit(EXIT_FAILURE);
			}

			vc_sasl_mech = ber_strdup(cvalue);
			if (vc_sasl_mech == NULL) {
				exit(EXIT_FAILURE);
			}
#else
#endif

		} else if (strcasecmp(control, "realm") == 0) {
#if defined(LDAP_API_FEATURES_VERIFY_CREDENTIALS_INTERACTIVE) && defined(HAVE_CYRUS_SASL)
			if (vc_sasl_realm) {
				fprintf(stderr,
				    _("SASL realm previously specified\n"));
				exit(EXIT_FAILURE);
			}
			if (cvalue == NULL) {
				fprintf(stderr,
					_("missing realm in SASL option\n"));
				exit(EXIT_FAILURE);
			}

			vc_sasl_realm = ber_strdup(cvalue);
			if (vc_sasl_realm == NULL) {
				exit(EXIT_FAILURE);
			}
#else
			fprintf(stderr,
				_("%s: not compiled with SASL support\n"), prog);
			exit(EXIT_FAILURE);
#endif

		} else if (strcasecmp(control, "authcid") == 0) {
#if defined(LDAP_API_FEATURES_VERIFY_CREDENTIALS_INTERACTIVE) && defined(HAVE_CYRUS_SASL)
			if (vc_sasl_authcid) {
				fprintf(stderr,
				    _("SASL authcid previously specified\n"));
				exit(EXIT_FAILURE);
			}
			if (cvalue == NULL) {
				fprintf(stderr,
					_("missing authcid in SASL option\n"));
				exit(EXIT_FAILURE);
			}

			vc_sasl_authcid = ber_strdup(cvalue);
			if (vc_sasl_authcid == NULL) {
				exit(EXIT_FAILURE);
			}
#else
			fprintf(stderr,
				_("%s: not compiled with SASL support\n"), prog);
			exit(EXIT_FAILURE);
#endif

		} else if (strcasecmp(control, "authzid") == 0) {
#if defined(LDAP_API_FEATURES_VERIFY_CREDENTIALS_INTERACTIVE) && defined(HAVE_CYRUS_SASL)
			if (vc_sasl_authzid) {
				fprintf(stderr,
				    _("SASL authzid previously specified\n"));
				exit(EXIT_FAILURE);
			}
			if (cvalue == NULL) {
				fprintf(stderr,
					_("missing authzid in SASL option\n"));
				exit(EXIT_FAILURE);
			}

			vc_sasl_authzid = ber_strdup(cvalue);
			if (vc_sasl_authzid == NULL) {
				exit(EXIT_FAILURE);
			}
#else
			fprintf(stderr,
				_("%s: not compiled with SASL support\n"), prog);
			exit(EXIT_FAILURE);
#endif

		} else if (strcasecmp(control, "secprops") == 0) {
#if defined(LDAP_API_FEATURES_VERIFY_CREDENTIALS_INTERACTIVE) && defined(HAVE_CYRUS_SASL)
			if (vc_sasl_secprops) {
				fprintf(stderr,
				    _("SASL secprops previously specified\n"));
				exit(EXIT_FAILURE);
			}
			if (cvalue == NULL) {
				fprintf(stderr,
					_("missing secprops in SASL option\n"));
				exit(EXIT_FAILURE);
			}

			vc_sasl_secprops = ber_strdup(cvalue);
			if (vc_sasl_secprops == NULL) {
				exit(EXIT_FAILURE);
			}
#else
			fprintf(stderr,
				_("%s: not compiled with SASL support\n"), prog);
			exit(EXIT_FAILURE);
#endif

		} else {
		    fprintf( stderr, _("Invalid Verify Credentials extension name: %s\n"), control );
		    usage();
		}
		break;

	case 'a':  /* request authzid */
		req_authzid++;
		break;

	case 'b':  /* request authzid */
		req_pp++;
		break;

	default:
		return 0;
	}
	return 1;
}


int
main( int argc, char *argv[] )
{
	int		rc;
	LDAP		*ld = NULL;
	char		*matcheddn = NULL, *text = NULL, **refs = NULL;
	int rcode;
	char * diag = NULL;
	struct berval	*scookie = NULL;
	struct berval	*scred = NULL;
	int		id, code = 0;
	LDAPMessage	*res;
	LDAPControl	**ctrls = NULL;
	LDAPControl	**vcctrls = NULL;
	int nvcctrls = 0;

	tool_init( TOOL_VC );
	prog = lutil_progname( "ldapvc", argc, argv );

	/* LDAPv3 only */
	protocol = LDAP_VERSION3;

	tool_args( argc, argv );

	if (argc - optind > 0) {
		dn = argv[optind++];
	}
	if (argc - optind > 0) {
		cred.bv_val = strdup(argv[optind++]);
		cred.bv_len = strlen(cred.bv_val);
	}
	if (argc - optind > 0) {
		usage();
	}
	if (dn 
#ifdef LDAP_API_FEATURE_VERIFY_CREDENTIALS_INTERACTIVE
           && !vc_sasl_mech 
#endif
           && !cred.bv_val)
	{
		char *userpw = getpassphrase(_("User's password: "));
		if ( userpw == NULL ) /* Allow EOF to exit. */
		{
			tool_exit( ld, EXIT_FAILURE );
		}
		cred.bv_val = strdup(userpw);
		cred.bv_len = strlen(cred.bv_val);
	}

#ifdef LDAP_API_FEATURE_VERIFY_CREDENTIALS_INTERACTIVE
    if (vc_sasl_mech && (vc_sasl == LDAP_SASL_NONE)) {
		vc_sasl = LDAP_SASL_AUTOMATIC;
	}
#endif

	ld = tool_conn_setup( 0, 0 );

	tool_bind( ld );

	if ( dont ) {
		rc = LDAP_SUCCESS;
		goto skip;
	}

	tool_server_controls( ld, NULL, 0 );

    if (req_authzid) {
		vcctrls = (LDAPControl **) malloc(3*sizeof(LDAPControl *));
		vcctrls[nvcctrls] = (LDAPControl *) malloc(sizeof(LDAPControl));
		vcctrls[nvcctrls]->ldctl_oid = ldap_strdup(LDAP_CONTROL_AUTHZID_REQUEST);
		vcctrls[nvcctrls]->ldctl_iscritical = 0;
		vcctrls[nvcctrls]->ldctl_value.bv_val = NULL;
		vcctrls[nvcctrls]->ldctl_value.bv_len = 0;
		vcctrls[++nvcctrls] = NULL;
    }

    if (req_pp) {
		if (!vcctrls) vcctrls = (LDAPControl **) malloc(3*sizeof(LDAPControl *));
		vcctrls[nvcctrls] = (LDAPControl *) malloc(sizeof(LDAPControl));
		vcctrls[nvcctrls]->ldctl_oid = ldap_strdup(LDAP_CONTROL_PASSWORDPOLICYREQUEST);
		vcctrls[nvcctrls]->ldctl_iscritical = 0;
		vcctrls[nvcctrls]->ldctl_value.bv_val = NULL;
		vcctrls[nvcctrls]->ldctl_value.bv_len = 0;
		vcctrls[++nvcctrls] = NULL;
    }

#ifdef LDAP_API_FEATURE_VERIFY_CREDENTIALS_INTERACTIVE
#ifdef HAVE_CYRUS_SASL
    if (vc_sasl_mech) {
		int msgid;
		void * defaults;
		void * context = NULL;
		const char *rmech = NULL;

		defaults = lutil_sasl_defaults(ld,
			vc_sasl_mech,
			vc_sasl_realm,
			vc_sasl_authcid,
			cred.bv_val,
			sasl_authz_id);

		do {
			rc = ldap_verify_credentials_interactive(ld, dn, vc_sasl_mech,
				vcctrls, NULL, NULL,
				vc_sasl, lutil_sasl_interact, defaults, context,
				res, &rmech, &msgid); 

			if (rc != LDAP_SASL_BIND_IN_PROGRESS) break;

			ldap_msgfree(res);

			if (ldap_result(ld, msgid, LDAP_MSG_ALL, NULL, &res) == -1 || !res) {
				rc = tool_perror2( ld, "ldap_verify_credentials_interactive" );
				tool_exit(ld, rc);
			}
		} while (rc == LDAP_SASL_BIND_IN_PROGRESS);

	    lutil_sasl_freedefs(defaults);

	    if( rc != LDAP_SUCCESS ) {
			ldap_get_option(ld, LDAP_OPT_DIAGNOSTIC_MESSAGE, (void*) &text);
		    tool_perror( "ldap_verify_credentials", rc, NULL, NULL, text, NULL );
		    rc = EXIT_FAILURE;
		    goto skip;
	    }

	} else
#endif
#endif
    {
	    rc = ldap_verify_credentials( ld,
		    NULL,
		    dn, NULL, cred.bv_val ? &cred: NULL, vcctrls,
		    NULL, NULL, &id ); 
    
	    if( rc != LDAP_SUCCESS ) {
			ldap_get_option(ld, LDAP_OPT_DIAGNOSTIC_MESSAGE, (void*) &text);
		    tool_perror( "ldap_verify_credentials", rc, NULL, NULL, text, NULL );
		    rc = EXIT_FAILURE;
		    goto skip;
	    }

	    for ( ; ; ) {
		    struct timeval	tv;

		    if ( tool_check_abandon( ld, id ) ) {
			    tool_exit( ld, LDAP_CANCELLED );
		    }

		    tv.tv_sec = 0;
		    tv.tv_usec = 100000;

		    rc = ldap_result( ld, LDAP_RES_ANY, LDAP_MSG_ALL, &tv, &res );
		    if ( rc < 0 ) {
			    rc = tool_perror2( ld, "ldap_result" );
			    tool_exit( ld, rc );
		    }

		    if ( rc != 0 ) {
			    break;
		    }
	    }
	}

	ldap_controls_free(vcctrls);
	vcctrls = NULL;

	rc = ldap_parse_result( ld, res,
		&code, &matcheddn, &text, &refs, &ctrls, 0 );

	if (rc == LDAP_SUCCESS) rc = code;

	if (rc != LDAP_SUCCESS) {
		tool_perror( "ldap_parse_result", rc, NULL, matcheddn, text, refs );
		rc = EXIT_FAILURE;
		goto skip;
	}

	rc = ldap_parse_verify_credentials( ld, res, &rcode, &diag, &scookie, &scred, &vcctrls );
	ldap_msgfree(res);

	if (rc != LDAP_SUCCESS) {
		tool_perror( "ldap_parse_verify_credentials", rc, NULL, NULL, NULL, NULL );
		rc = EXIT_FAILURE;
		goto skip;
	}

	if (rcode != LDAP_SUCCESS) {
		printf(_("Failed: %s (%d)\n"), ldap_err2string(rcode), rcode);
	}

	if (diag && *diag) {
	    printf(_("Diagnostic: %s\n"), diag);
	}

	if (vcctrls) {
		tool_print_ctrls( ld, vcctrls );
	}

skip:
	if ( verbose || code != LDAP_SUCCESS ||
		( matcheddn && *matcheddn ) || ( text && *text ) || refs || ctrls )
	{
		printf( _("Result: %s (%d)\n"), ldap_err2string( code ), code );

		if( text && *text ) {
			printf( _("Additional info: %s\n"), text );
		}

		if( matcheddn && *matcheddn ) {
			printf( _("Matched DN: %s\n"), matcheddn );
		}

		if( refs ) {
			int i;
			for( i=0; refs[i]; i++ ) {
				printf(_("Referral: %s\n"), refs[i] );
			}
		}

		if (ctrls) {
			tool_print_ctrls( ld, ctrls );
			ldap_controls_free( ctrls );
		}
	}

	ber_memfree( text );
	ber_memfree( matcheddn );
	ber_memvfree( (void **) refs );
	ber_bvfree( scookie );
	ber_bvfree( scred );
	ber_memfree( diag );
	free( cred.bv_val );

	/* disconnect from server */
	tool_exit( ld, code == LDAP_SUCCESS ? EXIT_SUCCESS : EXIT_FAILURE );
}
